tiledb/api/src/query/read/mod.rs

```diff
@@ -374,6 +375,104 @@ pub trait ReadQueryBuilder<'data>: QueryBuilder {
     }
 }
 
+pub struct CountReader<Q> {
+    base : Q
+}
+
+impl<Q> Query for CountReader<Q> where Q : Query {
+    fn base(&self) -> &QueryBase {
+        &self.base.base()
+    }
+
+    fn finalize(self) -> TileDBResult<Array>
+        where
+            Self: Sized {
+        self.base.finalize()
+    }
+}
+
+impl<Q> ReadQuery for CountReader<Q> where Q: ReadQuery, {
+    type Intermediate = ();
+    type Final = u64;
+
+    fn step(
+        &mut self,
+    ) -> TileDBResult<ReadStepOutput<Self::Intermediate, Self::Final>> {
+        // Register the data buffer (set data buffer)
+        let context = self.base().context();
+        let cquery = **self.base().cquery();
+        let location : *mut u64 = out_ptr!();
+        let size : *mut u64 = out_ptr!();
+        unsafe {
+            *size = 8;
+        }
+
+        let c_bufptr = location as *mut std::ffi::c_void;
+        let c_sizeptr = size as *mut u64;
+        let count_str = String::from("Count");
+        let count_c_ptr = count_str.as_ptr() as *const i8;
+
+        context.capi_call(|ctx| unsafe {
+            ffi::tiledb_query_set_data_buffer(
+                ctx,
+                cquery,
+                count_c_ptr,
+                c_bufptr,
+                c_sizeptr,
+            )
+        })?;
+
+        let base_result = self.base.step()?;
+        
+        // Run the query in a loop until you get the final result
+        let return_val = unsafe {match base_result {
+            ReadStepOutput::Final(_) => *location,
+            ReadStepOutput::Intermediate(_) => unreachable!(),
+            ReadStepOutput::NotEnoughSpace => unreachable!(),
+        }};
+
+        Ok(ReadStepOutput::Final(return_val))
+
+    }
+}
+
+pub enum AggregateType {
+    Count,
+    Sum
+    // agg_type : AggregateType , attr_name : Option<String>
+}
+
+pub trait AggregateBuilder : QueryBuilder {
```

Comment:
`/// Trait for query types which can have an aggregate channel placed on top of them`